test/SILOptimizer/moveonly_addresschecker_destructure_through_deinit_diagnostics.swift

```diff
@@ -153,3 +159,85 @@ func testStructContainStructContainDeinitStructConsumeNonCopyable4(_ x: consumin
+  // expected-note @-3 {{deinit declared here}}
+  // expected-note @-4 {{deinit declared here}}
+  // expected-note @-5 {{deinit declared here}}
+}
+
+func consume<T : P>(_ x: consuming AddressOnlyDeinitStruct<T>) {}
+
+func testAddressOnlyCanConsumeEntireType<T : P>(_ x: consuming AddressOnlyDeinitStruct<T>) {
+  // This is ok since we are consuming a copyable value.
+  consume(x.copyable)
+  // This is ok since we are consuming the entire value.
+  consume(x)
+}
+
+func testAddressOnlyCannotPartialConsume<T : P>(_ x: consuming AddressOnlyDeinitStruct<T>) {
+  // expected-error @-1 {{Cannot partially consume 'x' since it has a user defined deinit}}
+  // expected-error @-2 {{Cannot partially consume 'x' since it has a user defined deinit}}
+  consume(x.moveOnly) // expected-note {{consuming use here}}
+  consume(x.moveOnlyPair.first) // expected-note {{consuming use here}}
+  consume(x.copyable)
+}
+
+struct AddressOnlyContainingDeinitStruct<T : P> : ~Copyable {
+  var a = AddressOnlyDeinitStruct<T>()
+}
+
+func testAddressOnlyCannotPartialConsumeEvenIfSingleElt<T : P>(_ x: consuming AddressOnlyContainingDeinitStruct<T>) {
+  // expected-error @-1 {{Cannot partially consume 'x' since it contains field 'x.a' whose type 'AddressOnlyDeinitStruct' has a user defined deinit}}
+
+  // We do not error here since we can partially consume x, but not x.a
+  consume(x.a)
+  consume(x.a.moveOnlyPair) // expected-note {{consuming use here}}
+}
+
+struct AddressOnlyContainingDeinitSingleField<T : P> : ~Copyable {
+  var moveOnly = SingleIntContainingStruct()
+  deinit {}
+  // expected-note @-1 {{deinit declared here}}
+}
+
+struct AddressOnlyContainingDeinitStruct3<T : P> : ~Copyable {
+  var a = AddressOnlyContainingDeinitSingleField<T>()
+}
+
+func consume<T : P>(_ x: consuming AddressOnlyContainingDeinitSingleField<T>) {}
+
+func testAddressOnlyCannotPartialConsumeEvenIfSingleElt<T : P>(_ x: consuming AddressOnlyContainingDeinitStruct3<T>) {
+  // expected-error @-1 {{Cannot partially consume 'x' since it contains field 'x.a' whose type 'AddressOnlyContainingDeinitSingleField' has a user defined deinit}}
+
+  // We do not error here since we can partially consume x, but not x.a
+  consume(x.a)
+  consume(x.a.moveOnly) // expected-note {{consuming use here}}
+}
+
+
+struct AddressOnlyContainingDeinitStructPair<T : P> : ~Copyable {
+  var first = AddressOnlyDeinitStruct<T>()
+  var second = AddressOnlyDeinitStruct<T>()
+}
+
+// Make sure that if the deinit is in an intermediate element of the path that
+// we still handle it appropriately.
+func testAddressOnlyDeinitInMiddlePath<T : P>(_ x: consuming AddressOnlyContainingDeinitStructPair<T>) {
+  // expected-error @-1 {{Cannot partially consume 'x' since it contains field 'x.first' whose type 'AddressOnlyDeinitStruct' has a user defined deinit}}
+  // expected-error @-2 {{Cannot partially consume 'x' since it contains field 'x.first' whose type 'AddressOnlyDeinitStruct' has a user defined deinit}}
```

Comment:
This diagnostic is inaccurate, since `x` can be partially consumed if `x.first` or `x.second` is consumed whole. It would be better to use the same diagnostic as in the non-nested case, and say `cannot partially consume 'x.first' since it has a user defined deinit`.

Reply:
Ok. That makes sense.

Reply:
I'll make that diagnostic wording change in here: https://github.com/apple/swift/pull/66090

Reply:
@kavon. Ok. Thanks!